### PR TITLE
Use queued tasks in adaptive target

### LINKED ISSUE #8035: Adaptive target incorrectly assigns unknown duration to queued tasks
I'm running a workload right now that should be scaling up, but isn't.  I think I've tracked it down to these lines:

```python
        queued_occupancy = len(self.queued) * self.UNKNOWN_TASK_DURATION
        cpu = math.ceil(
            (self.total_occupancy + queued_occupancy) / target_duration
        )  # TODO: threads per worker
```

My tasks take several seconds, and I've run them before so the scheduler has a good estimate of task duration in the prefix.  Unfortunately this estimate isn't being used, instead `UNKNOWN_TASK_DURATION` is being used, and it's far less at about 0.5 s.

### PR DESCRIPTION
**Pull request by mrocklin**
Fixes #8035

### CODE REVIEW

**Comment by dchudz** [comment]
In case this helps Matt or reviewers: I tried out this PR and it makes my own go-to adaptive example go from scaling smaller than I expect to larger than I expect (because my workers have 4 threads each):


```
In [4]: c.adapt(minimum=1, maximum=200, target_duration="5 minutes")
2023-07-26 23:45:44,466 - distributed.deploy.adaptive - INFO - Adaptive scaling started: minimum=1 maximum=200
Out[4]: <coiled.cluster.CoiledAdaptive at 0x105725330>

In [5]: def sleep_5_secs(x):
   ...:     import time
   ...:     time.sleep(5)
   ...:     return x
   ...:

In [6]: results = client.map(sleep_5_secs, range(10_000))
```

My back-of-the-envelope says this should give me about 42 workers.  

Before this PR, it gave me 17. After this PR, it gives me 167, which is roughly 42*4, because adaptive doesn't account for worker threads:

```
        cpu = math.ceil(
            (self.total_occupancy + queued_occupancy) / target_duration
        )  # TODO: threads per worker
```

But I realize that's a separate problem that shouldn't block this PR. Just adding the context/example in case it helps.

**Comment by hendrikmakait** [comment]
It looks like `test_adaptive_target_empty_cluster[True]` fails consistently.

**Comment by mrocklin** [comment]
> It looks like test_adaptive_target_empty_cluster[True] fails consistently.

Thanks.  Fixed.

File: distributed/scheduler.py [hunk 1]
```diff
@@ -8062,8 +8062,26 @@ def adaptive_target(self, target_duration=None):
 
         # CPU
 
-        # TODO consider any user-specified default task durations for queued tasks
-        queued_occupancy = len(self.queued) * self.UNKNOWN_TASK_DURATION
+        if len(self.queued) < 100:
+            queued_occupancy = 0
+            for ts in self.queued:
+                if ts.prefix.duration_average == -1:
```
Comment (crusaderky) [inline distributed/scheduler.py:8068]: Out of scope for this PR: I think this is problematic. duration among the same TaskPrefix can vary wildly; I would much rather use a metric that is TaskGroup-specific.

File: distributed/scheduler.py [hunk 2]
```diff
@@ -8062,8 +8062,26 @@ def adaptive_target(self, target_duration=None):
...
+                    queued_occupancy += self.UNKNOWN_TASK_DURATION
+                else:
+                    queued_occupancy += ts.prefix.duration_average
```
Comment (crusaderky) [inline distributed/scheduler.py:8071]: Out of scope nit: this screams for encapsulation in a smart duration_average property

Comment (mrocklin) [reply distributed/scheduler.py:8071]: I don't disagree

**Review by crusaderky — CHANGES_REQUESTED** [review]
Both this PR and main fail (as the tests demonstrate) in the use case where

1. there are 0 workers, and
2. queuing is disable (`distributed.scheduler.worker-saturation: .inf`)

In this use case, all tasks will end up in `Scheduler.unrunnable` instead of `Scheduler.queued`.

Even when queueing is enabled, this fails when 
1. there are no workers
2. the tasks require resources

(I understand that the expectation in an adaptive cluster with resources is that all dynamically-started workers provide the resource, e.g. `{"GPU": 1}`).

Again, tasks will end up in `Scheduler.unrunnable`.
Please add a test for this use case.

File: distributed/scheduler.py [hunk 3]
```diff
@@ -8062,8 +8062,26 @@ def adaptive_target(self, target_duration=None):
...
+        else:
+            queued_occupancy = 0
+            queued = random.sample(self.queued._heap, 100)
+            queued = [wr() for _, _, wr in queued]
+            for ts in random.sample(queued, 100):
+                if ts is None:
+                    continue
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/scheduler.py:8078]: I think that random.sample is overkill here. The pseudo-randomness that is intrinsic in `HeapSet._data`, which is a set, should be enough.
So this should suffice, and doesn't require adding stuff to HeapSet:

```python
from toolz import take
queued = take(self.queued, 100) if len(self.queued) > 100 else self.queued
```

**Comment by crusaderky** [comment]
In short:
```python
from toolz import peekn

# Note: this relies on HeapSet.__iter__ to yield elements in pseudo-random order
queued, _ = peekn(100, self.queued)
if len(queued) < 100 and self.unrunnable:
    queued += peekn(100 - len(queued), self.unrunnable)[0]

queued_occupancy = 0
for ts in queued:
    if ts.prefix.duration_average == -1:
        queued_occupancy += self.UNKNOWN_TASK_DURATION
    else:
        queued_occupancy += ts.prefix.duration_average
```

**Comment by mrocklin** [comment]
As a heads-up I'm unlikely to spend a bunch of time on this.  It's more likely that I ask folks like @fjetter to ask people around him (maybe even @crusaderky ) to pick this up.

I'm hopeful that this can be a small fix.  I would be mildly sad/surprised if it required a large effort (not that that's what you're saying).

**Comment by mrocklin** [comment]
Never mind.  The use of take removes much of the concern about cost here (thanks for the suggestion).  

I do think that the request around supporting `unrunnable` is out of scope (this was broken before) but I've gone ahead and done it anyway.  I'm hopeful that we can get this in soon (I'm finding that this is valuable for coiled run work and this is blocking me).

File: distributed/deploy/tests/test_adaptive.py [hunk 4]
```diff
@@ -484,3 +484,22 @@ async def test_adaptive_stopped():
...
+
+@gen_cluster(
+    client=True,
+    nthreads=[],
+    config={"distributed.scheduler.default-task-durations": {"slowinc": 1000}},
+)
+async def test_scale_up_large_tasks(c, s):
+    futures = c.map(slowinc, range(10))
+    while not s.tasks:
+        await asyncio.sleep(0.001)
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/deploy/tests/test_adaptive.py:501]: I prefer not to use these.  I find that this removes one line but at the cost of adding a new abstraction (`async_poll_for`). The tradeoff here doesn't seem positive to me.

File: distributed/scheduler.py [hunk 5]
```diff
@@ -8062,8 +8062,26 @@ def adaptive_target(self, target_duration=None):
...
+                if ts.prefix.duration_average == -1:
+                    queued_occupancy += self.UNKNOWN_TASK_DURATION
+                else:
+                    queued_occupancy += ts.prefix.duration_average
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/scheduler.py]: I've done something like this.  Thank you for the suggestion.

File: distributed/scheduler.py (hunk 2 above)
Comment (fjetter) [reply distributed/scheduler.py]: There is [`get_task_duration`](https://github.com/dask/distributed/blob/9eb672840cc76f8718227fc1d2973cc6f233b9cd/distributed/scheduler.py#L2978-L2995) which also handles the case of user provided estimates.

**Review by fjetter — APPROVED** [review]

File: distributed/deploy/tests/test_adaptive.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/deploy/tests/test_adaptive.py:501]: We are using these ubiquitously. I think this is not a design choice that should be left to the whim and taste of the individual developers; if you don't like them we should have a team talk about them which should result in either using them everywhere or removing them completely.

**Pull request merged**